pkg/querier/worker/scheduler_processor.go

```diff
@@ -136,20 +137,45 @@ func (sp *schedulerProcessor) processQueriesOnSingleStream(workerCtx context.Con
 }
 
 // process loops processing requests on an established stream.
-func (sp *schedulerProcessor) querierLoop(c schedulerpb.SchedulerForQuerier_QuerierLoopClient, address string, inflightQuery *atomic.Bool) (err error) {
+func (sp *schedulerProcessor) querierLoop(execCtx context.Context, c schedulerpb.SchedulerForQuerier_QuerierLoopClient, address string, inflightQuery *atomic.Bool) (err error) {
 	// Build a child context so we can cancel a query when the stream is closed.
-	ctx, cancel := context.WithCancelCause(c.Context())
+	// Note that we deliberately don't use c.Context() here, as that is cancelled as soon as the gRPC client observes an error,
+	// but we don't always want to cancel queries if the scheduler stream reports an error (eg. if the scheduler crashed).
+	ctx, cancel := context.WithCancelCause(execCtx)
 	defer func() {
 		cancel(util.NewCancellationErrorf("query-scheduler loop in querier for query-scheduler %v terminated with error: %w", address, err))
 	}()
 
+	var queryComplete chan struct{}
+
+	waitForQuery := func(err error) {
+		select {
+		case <-queryComplete:
+			// Query is already complete, nothing to do.
+			return
+		default:
+			// Query is not complete.
+			level.Info(sp.log).Log("msg", "query-scheduler loop received non-cancellation error, waiting for inflight query to complete...", "err", err, "address", address)
+			<-queryComplete
+			level.Info(sp.log).Log("msg", "query-scheduler loop received non-cancellation error and inflight query is complete, continuing", "err", err, "address", address)
+		}
+	}
+
 	for {
 		request, err := c.Recv()
 		if err != nil {
+			// If the query was cancelled, we don't want to wait for it to complete: we want to cancel it, which will be
+			// handled by the deferred context cancellation above.
+			// If we got another kind of error (eg. scheduler crashed), continue processing the query.
+			if !grpcutil.IsCanceled(err) && queryComplete != nil {
```

Comment:
[nit] The `queryComplete == nil` case can only happen when the the `Recv()` returns error before the 1st query is executed (after that, `queryComplete` will be never nil). I don't feel strong, but I'm wondering if we should make it not a special case. An idea to not make it a special case is to initialise the `queryComplete` channel to a closed channel.

Reply:
Makes sense to me, changed in b5634d9